### ISSUE TITLE
Unmodifiabile result optimization

### ISSUE THREAD

**Issue by DanySK** [Protelis/Protelis#26]
We currently re-compute at each round the value for expressions that cannot change, e.g.:
1 + 1

For some kind of tree nodes, if we know that all the branches can not change, then we can mark them as "final" after the first evaluation, and stop recurring in unless an erasure operation happens.

@jakebeal and @mviroli, what is your opinion?

**Comment by jakebeal** [comment]
While I agree in principle, I see three issues that need consideration before we invest effort into this:

1. To what degree do such optimizations end up handled implicitly by Java's JITC?  I would not be surprised if it turns out that the relative benefit we can obtain is fairly small.
2. To what degree should this be handled at compile time (i.e., such that we can remove the branches altogether) vs. dynamically, as you propose?
3. How much are such operations actually costing programs, given that the highest cost operations tend to be external method calls or neighborhood calculations, neither of which is likely to see significant benefit from this type of optimization.

In short, I think this issue should remain open, but that pursuing it now is likely to add a great deal of complexity for little benefit - in short, the very definition of premature optimization.

**Comment by mviroli** [comment]
Standard Software Engineering says: don't do this unless you have 
evidence it really improves performance :)


On 04/08/2015 16:15, Jacob Beal wrote:
> While I agree in principle, I see three issues that need consideration
> before we invest effort into this:
>
>  1. To what degree do such optimizations end up handled implicitly by
>     Java's JITC? I would not be surprised if it turns out that the
>     relative benefit we can obtain is fairly small.
>  2. To what degree should this be handled at compile time (i.e., such
>     that we can remove the branches altogether) vs. dynamically, as you
>     propose?
>  3. How much are such operations actually costing programs, given that
>     the highest cost operations tend to be external method calls or
>     neighborhood calculations, neither of which is likely to see
>     significant benefit from this type of optimization.
>
> In short, I think this issue should remain open, but that pursuing it
> now is likely to add a great deal of complexity for little benefit - in
> short, the very definition of premature optimization.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Protelis/Protelis/issues/26#issuecomment-127627310>.
>

-- 
Dott. Ing. Mirko Viroli, PhD
Associate Professor
DISI, Alma Mater Studiorum - Università di Bologna, Cesena
via Sacchi 3, Tel.: +39 0547 338846
via Venezia 52, Tel.: +39 0547 339216
47521 Cesena (FC)
mailto:  mirko.viroli "at" unibo.it
http://apice.unibo.it/xwiki/bin/view/MirkoViroli/

**Comment by DanySK** [comment]
Let's just keep this open for future performance enhancements, if needed.

**Comment by DanySK** [comment]
@jakebeal 
1. Hard to state, I am not sure if a JIT can optimize at this level, I am no compiler expert :)
2. Optimizing at that stage may be better
3. We need profiling. I'd say not much, I agree that neighborhood calculations and reflections are currently our slowest operations. Method search has been optimized (through caching) recently, should be less of a problem now.

**Comment by jakebeal** [comment]
For methods, it's not even the reflection I'm thinking about, but the actual work done by the method that gets called (which we have zero control over).

**Comment by DanySK** [comment]
I'm answering myself here as I apparently got some more insights in the last four years.

1. Yes the JIT does inline and optimize at this level
2. Optimizing by entirely cutting branches away would be the only reasonable optimization, but its impact is likely negligible
3. I did a lot of Protelis profiling in these years :) . Not once this has been a performance hassle.

I'm glad you guys stopped me from applying premature optimization!
